### PR TITLE
Add a reusable function for filtering the list

### LINKED ISSUE #39: Add a reusable function for filtering the list
We currently have two components that comb through the list of items to find matches: 

* `ItemList.js` filters the list based on user input in the search field
* `AddItems.js` matches user input to the existing items to prevent duplicates 

It would be nice to extract this functionality into a new function that both components could use.

### PR DESCRIPTION
**Pull request by dariatsvetkova**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

We had two components that combed through the list of items to find matches:

* `ItemList.js` filters the list based on user input in the search field
* `AddItems.js` matches user input to the existing items to prevent duplicates

This PR is aimed to refactor the code used in both components by creating an external `filter` function that both components can use. The new function sanitizes user input and compares it to the array of existing items.

## Related Issue

Closes #39 

## Acceptance Criteria

* Functionality of `AddItems.js` and `ItemList.js` remains intact

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|   | :sparkles: New feature     |
|  ✓  | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Testing Steps / QA Criteria

To test the feature, use one of the following ways:

1) Follow the link to see the preview: 
2) Launch the app locally from your terminal:
* Pull down this branch with `git pull origin dt-filter-function` and check that branch out with `git checkout origin dt-filter-function`
* Run `npm start`  to launch the app in your browser at [http://localhost:3000](http://localhost:3000)

When on application:

* In Grocery List view, you should be able to filter the list based on your input.
* In Add Items view, when attempting to add an item that is already on your list, you should get an error message.
* In both views, case, punctuation and spaces at the beginning and end of the string will be ignored.
* In both views, entering special characters like `\` shouldn't cause an error.

### CODE REVIEW

File: src/ItemList.js [hunk 1]
```diff
@@ -12,13 +13,7 @@ function ItemList(props) {
...
   useEffect(() => {
     // [filter-list] 2. Comparison function to filter shopping list and create a search results array
-    // [filter-list] (Remove accidental space character from query, make lowercase, and prevent regex errors)
-    const resultsArray = props.list.filter((itemObj) => {
-      return itemObj['itemName']
-        .toLowerCase()
-        .replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
-        .includes(query.toLowerCase().trim());
-    });
+    const resultsArray = filter(props.list, query, false);
```
Comment (kkatoch89) [inline src/ItemList.js:16]: So much cleaner!

**Review by kkatoch89 — APPROVED** [review]
This is awesome @dariatsvetkova! Definitely helped clean the code! Thank you for doing this

**Review by luisaugusto — APPROVED** [review]
⭐️⭐️⭐️⭐️⭐️

File: src/Item.js [hunk 2]
```diff
@@ -4,12 +4,12 @@ import estimates from './lib/estimates';
...
 
 function Item({ userToken, item }) {
-  const { itemName, id, purchaseDates, purchaseEstimates  = [] } = item;
+  const { itemName, id, purchaseDates, purchaseEstimates = [] } = item;
   const [checked, setChecked] = useState(false);
 
   useEffect(() => {
     // Note that timeChecked value represents seconds (86400 secs in 24 hrs)
-    const timeChecked = 86400;
+    const timeChecked = localStorage.getItem('expirationDuration', 10) || 86400;
```
Comment (laurenmbeatty) [inline src/Item.js:12]: Is this leftover from the demo?

**Review by laurenmbeatty — APPROVED** [review]
V. v. nice!!!!! 🎉

File: src/lib/filter.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    return itemLettersFilt.join('');
+  };
+
+  const resultsArray = array.filter((itemObj) => {
+    if (exactMatch) {
+      // To prevent adding duplicate items
+      return checkItem(itemObj['itemName']) === checkItem(string);
+    } else {
+      // To search for an item in the list
+      return checkItem(itemObj['itemName']).includes(checkItem(string));
```
Comment (laurenmbeatty) [inline src/lib/filter.js:22]: These are good comments ^

**Review by cafloyd — APPROVED** [review]
GREAT work, Daria! You have such a good idea for optimizing code! This was an excellent idea

File: src/Item.js (hunk 2 above)
Comment (dariatsvetkova) [reply src/Item.js:12]: Yes, but I thought we might need it for testing in the future

File: src/lib/filter.js (hunk 3 above)
Comment (ablizben) [reply src/lib/filter.js:22]: I second Lauren! I actually had a question and wasn't paying attention to the comments and it was answered when I read them.

**Review by ablizben — APPROVED** [review]

**Pull request merged**